kong/runloop/handler.lua

```diff
@@ -760,9 +760,20 @@ return {
       if (protocols and protocols.https and not protocols.http and
           forwarded_proto ~= "https")
       then
-        header["connection"] = "Upgrade"
-        header["upgrade"]    = "TLS/1.2, HTTP/1.1"
-        return kong.response.exit(426, { message = "Please use HTTPS protocol" })
+        local redirect_status_code = route.https_redirect_status_code or 426
+
+        if redirect_status_code == 426 then
+          ngx.header["connection"] = "Upgrade"
+          ngx.header["upgrade"]    = "TLS/1.2, HTTP/1.1"
+          return kong.response.exit(426, { message = "Please use HTTPS protocol" })
+
+        elseif redirect_status_code == 301 or
+          redirect_status_code == 302 or
+          redirect_status_code == 307 or
+          redirect_status_code == 308 then
+          ngx.header["Location"] = "https://" .. forwarded_host .. var.request_uri
```

Comment:
Do we not handle non-standard ports? I.e. != 80 for http and != 443 for HTTPs.

Reply:
I'm not sure what you mean here.

Are you saying that the configuration should be able to redirect to HTTPS on a specific port rather than the default 443?

Reply:
> Are you saying that the configuration should be able to redirect to HTTPS on a specific port rather than the default 443?

Yes, if Kong's ssl proxy directive is listening on non-443 (or a previous proxy, as indicated by the forwarded port header). No?

Reply:
> Yes, if Kong's ssl proxy directive is listening on non-443 (or a previous proxy, as indicated by the forwarded port header). No?

It is possible that the proxy in-front of Kong is not the only proxy between the real client and Kong, in which case this might break, depending on if the proxy in front of Kong respects the forwarded port or not.  

Most common use case here is automated browser and clients which follow automated redirects. I have not seen any implementations which support or any issues pointing to the need for it yet (and if we have those, we can add it right now). 

We can add it if there comes an actual need for this in future.

Reply:
> depending on if the proxy in front of Kong respects the forwarded port or not.

Well, we know that is should respect it, so let's make the implementation safe for proxies that respect the specs, at least. If they don't, then it is not Kong's issue, but those proxies.

Currently this is broken not only for those proxies that respect the specifications, but also if Kong is listening on non-443 ports for HTTPS connections. It's a small change to make (along with an integration test) and I don't see a reason for not adding it _before_ shipping the feature, and having to come back to this further down the road.

Reply:
Okay, I went to see if this could be implemented and then I recalled having another config property (`https_redirect_port`) for this.

What happens is following: Client talks to LB or Kong via HTTP and then based on configuration, Kong decides to redirect.

Now, at this point, Kong is aware of the forwarded HTTP port that the LB is using or Kong is using but not about the SSL port.
To overcome this problem, in the beginning, I had introduced another property in the Route entity, which dictated the port to use for redirection.
I dropped that field as it seemed exposing too much configuration to the end user.

> Currently this is broken not only for those proxies that respect the specifications, but also if Kong is listening on non-443 ports for HTTPS connections.

We can use Kong's SSL port listener but then, we run into other problems:
- how do we handle the case when Kong doesn't have an SSL port and LB is doing TLS termination
- how do we handle the case when Kong has an ssl port of 8443 and the LB has 443 or vice-versa.

Reply:
> What happens is following: Client talks to LB or Kong via HTTP and then based on configuration, Kong decides to redirect.

That is the tricky case. In this case, I think it would be acceptable to assume that the LB should be sending the redirect on its own, before the request gets proxied through Kong in the first place. If we don't assume that then the property you suggested is relevant.

>     * how do we handle the case when Kong doesn't have an SSL port and LB is doing TLS termination

If Kong does not accept HTTPS connections but a Route is configured to be `https` only, that is certainly a larger issue that warrants warning logs and/or additional steps during initialization/router rebuilding. That said, it feels outside the immediate scope of this PR. Secondly, if the LB is doing TLS termination, then the requests to Kong will always be HTTP by definition, so this branch should not even be entered in the first place.

>     * how do we handle the case when Kong has an ssl port of 8443 and the LB has 443 or vice-versa.

In such cases, the terminating LB needs to set the forwarded port header, and Kong will consider it in priority when building the Location URL. In the "vice-versa" case, then Kong is the terminating LB, and any port accepting HTTPS should be fine (at least in Gateway mode), as long as there exists one.

Reply:
> If Kong does not accept HTTPS connections but a Route is configured to be https only, that is certainly a larger issue that warrants warning logs and/or additional steps during initialization/router rebuilding.

I might be wrong here but is it not the case that Kong will respect the forwarded proto header which will be `HTTP` or `HTTPS` and Kong will respect that while matching the route? So, it can happen that Kong routes traffic based on HTTP or HTTPS even if it itself is not configured to listen on a SSL port.


> In such cases, the terminating LB needs to set the forwarded port header, and Kong will consider it in priority when building the Location URL.

That is not possible because we want to use a different port in the redirect, right?
Let's say we have the following configuration:

LB (port 80 HTTP) -> Kong (port 8000 HTTP)
LB (port 443 HTTP) -> Kong (port 8443 HTTPS)

In such a case, when an HTTP request comes to LB and then to Kong, the forwarded port will be set to `80`. In case Kong decides to do a HTTPS redirect, it will want to use the SSL port of the LB, and not the HTTP port of the LB but Kong has no knowledge of that port while processing the request.

Reply:
> I might be wrong here but is it not the case that Kong will respect the forwarded proto header which will be HTTP or HTTPS and Kong will respect that while matching the route? So, it can happen that Kong routes traffic based on HTTP or HTTPS even if it itself is not configured to listen on a SSL port.

Correct! For trusted IPs at least... If none is configured (which is the case by default), then Kong will never be able to proxy those HTTPS-only Routes.

> In such a case, when an HTTP request comes to LB and then to Kong, the forwarded port will be set to 80. In case Kong decides to do a HTTPS redirect, it will want to use the SSL port of the LB, and not the HTTP port of the LB but Kong has no knowledge of that port while processing the request.

That is the case that warrants either assuming that the LB will perform the redirect on its own, or makes the case for your new `https_redirect_port` attribute (I wanted to highlight this in my previous comment but it got removed during editing I think).

We still don't cover the following case:

- `Client -> Kong (8000 HTTP + 8443 HTTPS)` need to redirect to `8443`.

Other cases I could think of:

- `Client -> LB (443) -> Kong (8000 HTTP, trusted)` LB forwarded proto, so the request will be considered "HTTPS".
- `Client -> LB (80) -> Kong (8000 HTTP, trusted)` LB should redirect, or Kong should have prior knowledge of the LB HTTPS port.
- `Client -> LB (443) -> Kong (8000 HTTP, not trusted)` LB forwarded proto, but isn't trusted. Misconfiguration.
- `Client -> LB (80) -> Kong (8000 HTTP, not trusted)` LB should redirect, or Kong should have prior knowledge of the LB HTTPS port.

Reply:
As mentioned during our brief f2f chat:

- You made the point that if a TCP LB is sitting in front of Kong, it becomes hard to determine whether Kong is the first hop in the server chain or not, and thus to smartly advertise a port to the client.
- On the `https_redirect_port` property: unless asked for, we feel it is outside of the scope of this PR and whether it will become needed at all or not. 

So for now, let's make the assumption that in order for HTTPS redirect to be effective, users must configure Kong to listen for HTTPS traffic on the standard 443 port. Once other minor concerns have been addressed here, let's merge this :+1: